src/ccl_halomod.c

```diff
@@ -0,0 +1,355 @@
+#include "gsl/gsl_sf_expint.h"
+#include "gsl/gsl_roots.h"
+#include "ccl_background.h"
+#include "ccl_power.h"
+#include "ccl_massfunc.h"
+#include "ccl_error.h"
+#include "ccl_halomod.h"
+
+// Analytic FT of NFW profile, from Cooray & Sheth (2002; Section 3 of https://arxiv.org/abs/astro-ph/0206508)
+// Normalised such that U(k=0)=1
+static double u_nfw_c(ccl_cosmology *cosmo, double c, double halomass, double k, double a, int *status){
+   
+  double rv, rs, ks, Dv;
+  double f1, f2, f3, fc;
+  double Delta_v = Dv_BryanNorman(cosmo, a, status); // Virial density of haloes
+
+  // Special case to prevent numerical problems if k=0,
+  // the result should be unity here because of the normalisation
+  if (k==0.) {    
+    return 1.;    
+  }
+
+  // The general k case
+  else{
+    
+    // Scale radius for NFW (rs=rv/c)
+    rv = r_delta(cosmo, halomass, a, Delta_v, status);
+    rs = rv/c;
+
+    // Dimensionless wave-number variable
+    ks = k*rs;
+
+    // Various bits for summing together to get final result
+    f1 = sin(ks)*(gsl_sf_Si(ks*(1.+c))-gsl_sf_Si(ks));
+    f2 = cos(ks)*(gsl_sf_Ci(ks*(1.+c))-gsl_sf_Ci(ks));
+    f3 = sin(c*ks)/(ks*(1.+c));
+    fc = log(1.+c)-c/(1.+c);
+  
+    return (f1+f2-f3)/fc;
+    
+  }
+}
+
+// Structure to help with calculating Bullock et al. (2001) formation time
+typedef struct{  
+  ccl_cosmology *cosmo;
+  double halomass;
+  int *status;
+} a_form_Bullock_func_Par;
+
+// Function for root solver for scale factor according to Bullock et al. (2001)
+static double a_form_Bullock_func(double a_form, void *params)
+{
+  a_form_Bullock_func_Par *p = (a_form_Bullock_func_Par *)params;;
+  double delta_c = 1.686;
+  return ccl_sigmaM(p->cosmo, p->halomass, a_form, p->status) - delta_c;  
+}
+
+// Function to return the formation scale factor according to Bullock et al. (2001)
+double a_form_Bullock(ccl_cosmology *cosmo, double halomass, double a, int *status)
+{
+  int gslstatus;
+  int iter = 0, max_iter = 100;
+  const gsl_root_fsolver_type *T;
+  gsl_root_fsolver *s;
+
+  double a_form = 0;
+  double a_max = 1.0, a_min = 1./(1.+1000.0);
+  gsl_function F;
+
+  a_form_Bullock_func_Par fpar;
+
+  fpar.cosmo = cosmo;
+  fpar.halomass = halomass;
+  fpar.status = &gslstatus;
+
+  F.function = &a_form_Bullock_func;
+  F.params = &fpar;
+
+  T = gsl_root_fsolver_brent;
+  s = gsl_root_fsolver_alloc (T);
+  gsl_root_fsolver_set (s, &F, a_min, a_max);
+
+  do
+    {
+    iter++;
+    gslstatus = gsl_root_fsolver_iterate (s);
+    a_form = gsl_root_fsolver_root (s);
+    a_min = gsl_root_fsolver_x_lower (s);
+    a_max = gsl_root_fsolver_x_upper (s);
+    gslstatus = gsl_root_test_interval (a_min, a_max, 0., 0.001);
+
+    // testing print statements until it actually works
+    printf("%5d [%.7f, %.7f] %.7f", iter, a_min, a_max, a_form);
+    }
+  while (gslstatus == GSL_CONTINUE && iter < max_iter);
+
+  gsl_root_fsolver_free (s);
+
+  return a_form;    
+}
+
+// The concentration-mass relation for haloes
```

Comment:
Since this is a `ccl_` function (i.e. user-facing), it should have a proper docstring comment.

Reply:
Done